abci/check_tx.go

```diff
@@ -0,0 +1,186 @@
+
+	// CheckTxMempool is the interface that defines all of the dependencies that
+	// are required to interact with the application-side mempool.
+	CheckTxMempool interface {
+		// GetAuctionBidInfo is utilized to retrieve the bid info of a transaction.
+		GetAuctionBidInfo(tx sdk.Tx) (*mempool.AuctionBidInfo, error)
+
+		// Insert is utilized to insert a transaction into the application-side mempool.
+		Insert(ctx context.Context, tx sdk.Tx) error
+
+		// WrapBundleTransaction is utilized to wrap a transaction included in a bid transaction
+		// into an sdk.Tx.
+		WrapBundleTransaction(tx []byte) (sdk.Tx, error)
+	}
+
+	// BaseApp is an interface that allows us to call baseapp's CheckTx method
+	// as well as retrieve the latest committed state.
+	BaseApp interface {
+		// CommitMultiStore is utilized to retrieve the latest committed state.
+		CommitMultiStore() sdk.CommitMultiStore
+
+		// CheckTx is baseapp's CheckTx method that checks the validity of a
+		// transaction.
+		CheckTx(cometabci.RequestCheckTx) cometabci.ResponseCheckTx
+
+		// Logger is utilized to log errors.
+		Logger() log.Logger
+
+		// LastBlockHeight is utilized to retrieve the latest block height.
+		LastBlockHeight() int64
+	}
+)
+
+// NewCheckTxHandler is a constructor for CheckTxHandler.
+func NewCheckTxHandler(baseApp BaseApp, txDecoder sdk.TxDecoder, mempool CheckTxMempool, anteHandler sdk.AnteHandler, chainID string) CheckTxHandler {
+	return CheckTxHandler{
+		baseApp:     baseApp,
+		txDecoder:   txDecoder,
+		mempool:     mempool,
+		anteHandler: anteHandler,
+		chainID:     chainID,
+	}
+}
+
+// CheckTxHandler is a wrapper around baseapp's CheckTx method that allows us to
+// verify bid transactions against the latest committed state. All other transactions
+// are executed normally. We must verify each bid tx and all of its bundled transactions
+// before we can insert it into the mempool against the latest commit state because
+// otherwise the auction can be griefed. No state changes are applied to the state
+// during this process.
+func (handler *CheckTxHandler) CheckTx() CheckTx {
+	return func(req cometabci.RequestCheckTx) (resp cometabci.ResponseCheckTx) {
+		defer func() {
+			if err := recover(); err != nil {
+				resp = sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("panic in check tx handler: %s", err), 0, 0, nil, false)
+			}
+		}()
+
+		sdkTx, err := handler.txDecoder(req.Tx)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("failed to decode tx: %w", err), 0, 0, nil, false)
+		}
+
+		// Attempt to get the bid info of the transaction.
+		bidInfo, err := handler.mempool.GetAuctionBidInfo(sdkTx)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("failed to get auction bid info: %w", err), 0, 0, nil, false)
+		}
+
+		// If this is not a bid transaction, we just execute it normally.
+		if bidInfo == nil {
+			return handler.baseApp.CheckTx(req)
+		}
```

Comment:
Why have an explicit check for this? Wouldn't the flow be simpler to just explicitly verify `bidInfo.Transactions` and then at the end, run `return handler.baseApp.CheckTx(req)`?

Reply:
In general, all transactions that are not bid transactions should just execute base app's check tx. We could verify the bid transaction and all of the bundled transactions on top of the latest commit state first and then run `handler.baseApp.CheckTx(req)` but that might bring up other issues since we are now validating the bid tx on `CheckTx` state which may lead to unexpected differences in whether a transaction is valid.